View Factor Discrepancy when Using Autosize for Ground View Factor for Window with Light Shelf

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8154 
 - When a user selects autosize for the view factor to ground for a window and then has a light shelf attached to the window, the sum of the three view factors used for the window (view factor to ground, sky, and shelf) add up to greater than zero prompting a warning message.  While the warning message is helpful, it doesn't actually do anything to rectify the situation and continues the simulation with inappropriate view factors.  This is now clarified and the view factors will be fixed.  A warning message is still produced to let the user know what is happening.  While the documentation discourages using autosize for situations like this one, an additional note was put in the documentation so that users know more specifically the cause of the error message.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
8154 Fix for Autocalculate View Factor with Light Shelf

This contains the fix for the issue and the unit test.  When using autosizing for the view factors of a window that also has a light shelf, the view factors to ground and sky were set to values based on tilt.  These two values ended up being equal to 1.0 when combined.  However, then the light shelf calculated a view factor and added that to the other values resulting in the sum of the view factors being greater than zero.  This is now fixed.

RKStrand committed:
8154 Lightshelf View Factor Fix Format and Docs

Fixed a clang-format issue and also added a new note in the documentation about this issue.

RKStrand committed:
Merge branch 'develop' into 8154AutoCalcViewFactorDetailedFenestrationwithLightShelf

jcyuan2020 reviewed (commented):
A general thought about the methodology: it might be a little bit more complicated based on the geometric relation with window. For example, if the light shelf is aligned with the bottom of the window, then only the view factor to the ground should be reduced (by subtracting the amount of the view factor to the light-shelf), rather than taken proportionally off from both the sky and the ground part. 

I checked the example file DaylightingDeviceShelf.idf. It looks like that is a example case where the light shelf is aligned at the bottom of the "upper window". It then seems for this upper window, the view factor to the ground would be adjusted as to 0.5-0.29=0.21 and the then view factor to the sky is still kept at 0.5. 

Actually I wonder if this should belong to some automatic calculation part of the program when figuring out the view factor based on the surfaces' geometric relationships?

RKStrand commented:
@jcyuan2020 That's a good question.  I agree that the fix that I posted is an approximation.  I'm not sure at this point how much work it will take to make the correction based on the actual geometry, but I will look into that and report here.

RKStrand commented:
@jcyuan2020 I've looked at this a bit more closely.  The CalcViewFactorToShelf method simply assumes that the light shelf is not in the middle of the window but shares two vertices.  The assumption here is that the light shelf is either at the top, bottom, or side of the window.  While the side seems unlikely, EnergyPlus does not really test for that or really care.  I think I can trap for where the light shelf is on the window by looking at the z-coordinate of the matched vertices and comparing them to the window.  If the z-coordinates for the light shelf are the same and along the lower edge of the window, then I can use the calculated view factor to the shelf to reduce the view factor to ground by whatever the shelf view factor is.  If along the upper edge, reduce the view factor to the sky.  If not along the upper or lower edge, the use the new distribution reduction that I programmed it.  I think that would address your concern and be the best that we could do without something a lot more sophisticated.  I'm going to move in that direction.  However, if that still doesn't address your concerns, please let me know.  Thanks.

jcyuan2020 commented:
@RKStrand Sorry that my thought comment might have hindered the progress on this PR. I was just trying to toss some ideas and I do feel that a neat solution to cover all scenarios might not be easy (if not possible at all). I gave the question a little bit more thought and think probably we may excise some assumptions to simply (just an ideal, still). I looked at the introduction part about the light shelves in EngineeringRef and found the following figure. 
![image](https://user-images.githubusercontent.com/64040111/132064362-a4eef52c-aa9a-4f40-a25a-e4de15ae17e8.png)

There seems to be some general assumption that the shelf is pretty much aligned with the bottom of the window (the IORef mentioned it should consider the upper and lower window separately--so hopefully this will convert the upper window to a light-shelf problem and the lower window to a shading problem). In this case, the sky view factor would not be affected---this does make sense to the light shelf design since its self would not block a chunk of visible sky to the upper window its usage is to enhance light entering from the upper window. For a windows that does not sees only the grounds (and sky and the shelf) but not the building itself, the reduction would be fully from the ground. Actually this is the case in the example file. 

Probably at least to scenarios would complicate the issue, for example: 

1). One possible complicated case might be that the shelf got--actually I am still checking if this is possible from the IORef specifications. There might be a way, but would involve too much geometric/surface manipulation to divide the shelf surface into a part below and a part above the window bottom. 

2). The other complication, even for the simple bottom-flush case is when the upper window can see the building exterior surfaces itself (e.g a concave building), the blocked view by the shelf then might belongs to the building surface and then that projected part should not taken away from the ground or the sky. This is complicated in that there are quite a few different combinations on where and how much to take away from each part. For example, the bottom aligned case, then the view factor should taken away the part of view factor from the ground, but only for the part that not intercepted by the building surface itself.

BTW, I think the CalcViewFactor() function is fine--it will probably just use some method calculate the view factor of two polygon surfaces--and it should handle the window-to-shelve without any problems. The question is how we use the calculated view factor to the shelf, (and combine the shelf location and building surfaces) to get the proper reduction to  the view-to-sky and the view-to-ground part.

RKStrand commented:
@jcyuan2020 Ok, so if the shelf is supposed to be on the bottom of the window, then perhaps, to summarize, what we need to do is not what I originally proposed, but rather to simply reduce the view to ground by what is calculated as the view to the light shelf?  That would be pretty straightforward and I think it would be pretty reasonable.  Do you agree with that as a viable solution?  I think there probably are quite more complicated solutions, but I think that would probably need to be a new feature that investigates this more thoroughly.  The point of this bug fix (as I understood it) seems to be to make sure that we aren't having view factors that add up to greater than 1.  Thoughts?

jcyuan2020 commented:
@RKStrand I think that is definitely reasonable enough. I took a look at the input for the IORef about the window-shelf part, and it explicitly mentioned that the windows need to be divided into an upper one and lower one. Then probably having a bottom aligned shelf that would not obtrude the sky view (e.g. a horizontal one) might be the most common way to use a light shelf--so here probably most common case would be covered and use the to-ground reduction only. 

The IORef mentioned that the shelf is input via a Shading:Zone:Detailed object, which theoretically means it can be more complicated surface--e.g. has the extend-out edge to have higher position than the window---so here the sky view reduction might be needed. The criterion is that the shelf has at least one vertex having a higher z-axis value than the bottom of the window z-axis value. To limit scope, under this scenario, it could be either giving some special warning to let the user to consider this effect manually; or probably using your original solution of distributing the to-sky and to-ground reduction for this case--but also at the same time giving some warnings about the potential accuracy/inaccuracy of the adjustment. 

These are just some thoughts. Feel free to adjust the scope as you see fit.

RKStrand committed:
Merge branch 'develop' into 8154AutoCalcViewFactorDetailedFenestrationwithLightShelf

jcyuan2020 commented:
@RKStrand Some thoughts about a possible logic might look like this: 
1. Find the lowest vertex z-value (e.g., zmin_win) in the window surface, by looping the vertices in  state.dataSurface->Surface(WinSurf).Vertex 
2. Find the highest vertex z-value (zmax_shelf) in the light shelf surface, by looping the vertices state.dataSurface->Surface(ShelfSurf).Vertex
3. if (zmin_win >= zmax_shelf) { 
        //  no skyview-blocking case, probably very common
       // deduct from the ground-view factor only;
       //  and give some warnings that adjustment occurred
    } else { // if zmin_win < zmax_shelf 
       // warn the user the shelf is blocking both part of skyview and groundview
       // either ask the user to check or intervene manually 
      // or apply the sky and ground reduction based some proportions and give some extra warnings
   }

There could be another `else if` branch in between to consider the case of zmax_window <= zmin_shelf, but this might be a rare case for light shelf (light shelf aligned or higher than the window top).

RKStrand committed:
8154 Corrections based on Review

Correction to methodology based on review from @jcyuan2020.  View factor to ground reduced first followed by view factor to sky since standard placement of shelf is on the bottom edge of the upper window.

RKStrand committed:
8154 Doc Fix based on Review

Minor docs fix to clarify the method.

RKStrand commented:
@jcyuan2020 I've committed a new version of the method and a minor doc modification.  This branch was also brought up to date with develop.  The new method takes away from view to ground first and then view to sky.  Hopefully CI comes back clean and this is good to go.

RKStrand commented:
@jcyuan2020 Our most recent comments somewhat crossed paths.  I will look into what you proposed in your last comment and see what I think.

RKStrand committed:
8154 Changes to View Factor after Review Comments

As a result of comments made by @jcyuan2020, the method for figuring out view factors to ground and sky when the calculated view factor to shelf causes the sum of all three to be greater than zero was completely changed.  The method is now more robust and follows the suggestions with some additional features as well.

RKStrand committed:
Merge branch 'develop' into 8154AutoCalcViewFactorDetailedFenestrationwithLightShelf

RKStrand commented:
@jcyuan2020 Ok, I have rewritten this routine--it's much more detailed now and hopefully more robust.  I have used your suggestion as a starting point and then developed further from there.  There are cases for when the shelf is below the window, above the window, and also somewhere in between.  Also rewrote the unit test based on this new method.  I have committed all of this as well as merged in the latest develop.  Hopefully this looks good now and CI comes back clean.  Please let me know if you have any other concerns.  Thanks!

RKStrand committed:
Merge branch 'develop' into 8154AutoCalcViewFactorDetailedFenestrationwithLightShelf

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+                heightRatio = max(heightRatio, 0.0);
+            } else { // Window does not have a positive height (not realistic) so set height ratio based on shelf location
+                if (zShelfAvg > zWinMax) {
+                    heightRatio = 1.0;
```
The inner `if` and `else` are getting the same results--so just one line `heightRatio = 1.0;` should suffice.

RKStrand replied:
@jcyuan2020 I'm not sure I understand your comment here.  Starting in line 1742, we have to have some window height to avoid a divide by zero in heightRatio.  The else that starts at 1746 should never happen but I'm just making sure the case is covered.  Are you saying lines 1747 through 1751 should just be replaced by heightRatio = 1.0?  That just checks where the shelf is with respect to the window.  I guess it's not necessary, but I want to make sure I understand what you are saying correctly so I get this right.

jcyuan2020 replied:
Sorry @RKStrand , my bad! I thought the line in the `else` bracket was also `=1.0` when browsing through it initially.

RKStrand replied:
No problem--the code here is a little tedious and it's easy to miss that detail.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+                heightRatio = (zShelfAvg - zWinMin) / (zWinMax - zWinMin);
+                heightRatio = min(heightRatio, 1.0);
+                heightRatio = max(heightRatio, 0.0);
+            } else { // Window does not have a positive height (not realistic) so set height ratio based on shelf location
```
This `else` scenario can become "activated" when it is a horizontal window--but anyway a weird scenario to work with light shelf.

RKStrand replied:
@jcyuan2020 Yes, this would be a weird scenario...I was trying to be as complete as possible, but it does have a built-in assumption that this is being used for a vertical wall.  I'm not sure that I have ever seen a lightshelf used with a skylight before.

jcyuan2020 replied:
No problem---just trying to imagine a scenario where this block could be executed.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+            }
+
+        } else if (zShelfMin >= zWinMax) { // Shelf is fully above window, reduce view to sky first based on view to shelf
+            ShowContinueError(
```
It is nice to consider this scenario. Hope this scenario would never happen in real light shelf design. Otherwise, it is becomes the shading problem that is the other half of the light-shelf physics---where the lower window is modeled with shading phenomena.

RKStrand replied:
Thanks--I was trying to be as thorough as possible.  I think what we ended up with is much better than what I had initially and more fully addresses the problem.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+            ShowContinueError(
+                state,
+                "to the ground and sky were both potentially reduced Check you input and/or consider turning off autosizing of the view factors.");
+            Real64 zShelfAvg;
```
Missing a period `.` after the first sentence (after the word "reduced").

RKStrand replied:
Nice catch--this should be fixed in the next commit.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+        int ShelfSurf = state.dataDaylightingDevicesData->Shelf(ShelfNum).OutSurf;
+        Real64 zShelfMax = -9999.9;
+        Real64 zShelfMin = 9999.9;
+        for (int vertex = 1; vertex <= 4; ++vertex) {
```
state.dataDaylightingDevicesData->Shelf(ShelfNum).Sides will get the number of vertices of the surface, which can be used to replace the `4` to allow more general polygon.

RKStrand replied:
Yes, I agree.  Change will show up in the next commit.  Will need to also change the unit test to set this correctly.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+        // If the flow is still here, there is something that needs to be adjusted so set the maximum shelf height and the minimum window height
+        int ShelfSurf = state.dataDaylightingDevicesData->Shelf(ShelfNum).OutSurf;
+        Real64 zShelfMax = -9999.9;
+        Real64 zShelfMin = 9999.9;
```
`zShelfMax` and `zShelfMin` should be initialized to `state.dataDaylightingDevicesData->Shelf(ShelfNum).Vertex(0).z` before the loops starts, to avoid a rare case where 9999 or -9999 got returned instead of one of the vertices' z-ordinate.

RKStrand replied:
While this would be pretty rare given the values set here, I agree that it makes more sense to use the first vertex to set the Max and Min values and then go from there.  This will be changed in the next commit.

jcyuan2020 replied:
@RKStrand Sorry here it should be `state.dataDaylightingDevicesData->Shelf(ShelfNum).Vertex(1).z` as the vertex index start with 1 instead of zero.

RKStrand replied:
No problem.  I knew what you meant.  state.dataSurface->Surface(ShelfSurf).Vertex(1).z works in this case.  I went with that.

jcyuan2020 replied:
Also, then the loop can start with `for(int vertex = 2;...`---the first one no longer needs to be compared.

RKStrand replied:
Yes.  I caught that.  All of the changes have now been committed.  Hoping for a clean CI and that all of the concerns have been addressed now.  Thanks for your review!

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+                zShelfMin = state.dataSurface->Surface(ShelfSurf).Vertex(vertex).z;
+        }
+        Real64 zWinMax = -9999.9;
+        Real64 zWinMin = 9999.9;
```
Same here for initialization.

RKStrand replied:
Ditto here--fixed in the next commit.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+        Real64 zWinMax = -9999.9;
+        Real64 zWinMin = 9999.9;
+        for (int vertex = 1; vertex <= 4; ++vertex) {
+            if (state.dataSurface->Surface(WinSurf).Vertex(vertex).z > zWinMax) zWinMax = state.dataSurface->Surface(WinSurf).Vertex(vertex).z;
```
Same here to use `state.dataSurface->Surface(WinSurf).Sides` instead of '4'.

RKStrand replied:
Yes, I agree.  Change will show up in the next commit.  Will need to also change the unit test to set this correctly.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+                    viewFactorToGround = 0.0;
+                    viewFactorToShelf = 1.0;
+                }
+            }
```
So for after all done for this case, maybe a little bit extra (optional) warning the user should still check the view from the window to the building itself which might complicate the case and use manual input when applicable(?).

RKStrand replied:
I will add another warning at the end of this routine that specifically mentions this and covers all of the cases where something had to be reduced.

jcyuan2020 commented on src/EnergyPlus/DaylightingDevices.cc:
```diff
@@ -1620,6 +1614,164 @@ namespace DaylightingDevices {
+            viewFactorToGround = vfGroundAdjustMin + heightRatio * (vfGroundAdjustMax - vfGroundAdjustMin);
+            viewFactorToSky = leftoverViewFactor - viewFactorToGround;
+        }
+    }
```
After all done for this case, maybe a little bit extra (optional) warning that the estimate may not be a precise calculation; (and also there could be scenarios where the view from the window to the building itself which might complicate the case as well)?

RKStrand replied:
See previous comment.  I'll put one additional warning at the end of the routine since if it gets there it's because there was something reduced and the user really needs to think through things a bit more.

jcyuan2020 reviewed (commented):
@RKStrand Thanks for the updates! The algorithm nicely considers quite a few different scenarios now. A few comments were inline with the code.

RKStrand committed:
8154 Light Shelf View Factor Addressed Review Comments

Another round of comments on this defect were addressed.  Hopefully a final version of this work.  Significantly improved by the review process and involvement of @jcyuan2020.

RKStrand commented:
@jcyuan2020 The latest changes to address all of your review comments have been committed and CI is coming back all green.  Is this ready for merging in your opinion?

jcyuan2020 commented:
@RKStrand I tested a few scenarios (e.g. the regular flat bottom-aligned and one with modification a slightly tilted up shelf) and the results look  good and warning messages are adequately. 
I had a full local CI run a couple of days ago on a PC but the run was interrupted by an electricity blackout by the hurricane--actually we are still in a blackout now.  But since the web CI results were clean and I also just finished a selected random 30 CI runs on a laptop and it came back clean as well. I think this one should be good to go.

mjwitte commented:
Lots of code here to improve the default view factors for windows with a light shelf. Thanks @RKStrand and thanks @jcyuan2020 for the review. Merging.
